Translate "Component State" to German (#101)
reactjs/de.reactjs.org: content/docs/faq-state.md

---
id: faq-state

> title: Zustand von Komponenten

Reviewer: ```suggestion
title: State in Komponenten
```

We decided to write "State", because "Zustand" is disheveling. (setState, this.state etc.)

Author: OK, did not know about that decision.

Reviewer: All good :) We try to write down all these words (https://github.com/reactjs/de.reactjs.org/issues/10)

permalink: docs/faq-state.html
layout: docs
category: FAQ
---

### Was macht `setState`? {#what-does-setstate-do}

`setState()` veranlasst eine Änderung am `state`-Objekt einer Komponente. Sobald sich das Objekt ändert, reagiert die Komponente mit erneutem Rendern.

### Was ist der Unterschied zwischen `state` und `props`? {#what-is-the-difference-between-state-and-props}

> [`props`](/docs/components-and-props.html) (kurz für englisch "properties") und [`state`](/docs/state-and-lifecycle.html) sind beide einfache JavaScript-Objekte. Obwohl beide Informationen halten, die das Erebnis des Renderns beeinflussen, unterscheiden sie sich in einem entscheidenden Punkt: `props` werden *an* die Komponente gereicht (ähnlich zu Funktionsparametern), während `state` *in* der Komponente verwaltet wird (ähnlich zu Variablen, die in einer Funktion deklariert werden).

Reviewer: ```suggestion
[`props`](/docs/components-and-props.html) (kurz für englisch "properties") und [`state`](/docs/state-and-lifecycle.html) sind beide einfache JavaScript-Objekte. Obwohl beide Informationen halten, die das Ergebnis des Renderns beeinflussen, unterscheiden sie sich in einem entscheidenden Punkt: `props` werden *an* die Komponente gereicht (ähnlich zu Funktionsparametern), während `state` *in* der Komponente verwaltet wird (ähnlich zu Variablen, die in einer Funktion deklariert werden).
```

Hier sind einige gute Quellen dazu, wann man `props` verwenden sollte und wann `state`:
* [Props vs State](https://github.com/uberVU/react-guide/blob/master/props-vs-state.md)
* [ReactJS: Props vs. State](https://lucybain.com/blog/2016/react-state-vs-pros/)

### Warum gibt `setState` mir einen falschen Wert? {#why-is-setstate-giving-me-the-wrong-value}

In React repräsentieren sowohl `this.props` als auch `this.state` den *gerenderten* Wert, d.h. das was aktuell angezeigt wird.

Aufrufe an  `setState` sind asynchron - verlasse dich nicht darauf, dass `this.state` direkt nach einem Aufruf an `setState` den neuen Wert repräsentiert. Verwende eine Updater-Funktion statt einem Objekt, wenn du Werte basierend auf dem aktuellen Zustand berechnen musst (siehe folgendes Beispiel für Details).

Beispiel-Code, der sich *nicht* wie erwartet verhält:

```jsx

> erhoeheZaehler() {

Reviewer: ```suggestion
incrementCount() {
```
We have decided to leave method names as they are and only translate text.

Author: Done in 6904e2b2b32810264be3cc0c6bb47f5702f90fcf, resolving all comments now.

  // Achtung: dies verhält sich *nicht* wie erwartet.

>   this.setState({count: this.state.zaehler + 1});

Reviewer: ```suggestion
  this.setState({count: this.state.count + 1});
```

}

> behandleEtwas() {

Reviewer: ```suggestion
handleSomething() {
```

>   // Nehmen wir an, `this.state.zaehler` startet bei 0.

Reviewer: ```suggestion
  // Nehmen wir an, `this.state.count` startet bei 0.
```

>   this.erhoeheZaehler();

Reviewer: ```suggestion
  this.incrementCount();
```

>   this.erhoeheZaehler();

Reviewer: ```suggestion
  this.incrementCount();
```

>   this.erhoeheZaehler();

Reviewer: ```suggestion
  this.incrementCount();
```

>   // Wenn React die Komponente neu rendert, wird `this.state.zaehler` 1 sein, obwohl du 3 erwartest.

Reviewer: ```suggestion
  // Wenn React die Komponente neu rendert, wird `this.state.count` 1 sein, obwohl du 3 erwartest.
```

>   // Das liegt daran, dass `erhoeheZaehler()` von `this.state.zaehler` liest,

Reviewer: ```suggestion
  // Das liegt daran, dass `incrementCount()` von `this.state.count` liest,
```

>   // aber React `this.state.zaehler` nicht verändert bevor die Komponente neu gerendert wird.

Reviewer: ```suggestion
  // aber React `this.state.count` nicht verändert bevor die Komponente neu gerendert wird.
```

>   // Daher liest `erhoeheZaehler()` jedes Mal den Wert 0 von `this.state.zaehler`,

Reviewer: ```suggestion
  // Daher liest `incrementCount()` jedes Mal den Wert 0 von `this.state.count`,
```

  // und setzt ihn auf 1.

  // Die richtige Lösung findet sich weiter unten!
}
```

Siehe im Folgenden, wie man das Problem beheben kann.

> ### Wie kann ich den Zustand verändern mit Werten, die vom aktuellen Zustand abhängen? {#how-do-i-update-state-with-values-that-depend-on-the-current-state}

Reviewer: ```suggestion
### Wie kann ich den State verändern mit Werten, die vom aktuellen State abhängen? {#how-do-i-update-state-with-values-that-depend-on-the-current-state}
```

> Gib eine Funktion statt einem Objekt an `setState` um sicherzustellen, dass der Aufruf immer den aktuellsten Wert des Zustands verwendet (siehe unten). 

Reviewer: ```suggestion
Gib eine Funktion statt einem Objekt an `setState` um sicherzustellen, dass der Aufruf immer den aktuellsten Wert des States verwendet (siehe unten). 
```

### Was ist der Unterschied, wenn man ein Objekt oder eine Funktion an `setState` gibt? {#what-is-the-difference-between-passing-an-object-or-a-function-in-setstate}

> Gibt man eine Funktion an `setState`, so kann man in der Funktion auf den aktuellen Wert des Zustands zugreifen. Nachdem Aufrufe an `setState` gebündelt durchgeführt werden, lässt dich das Veränderungen aufeinanderfolgend durchführen und du kannst dir sicher sein, dass die Veränderungen aufeinander aufbauen anstatt zu konkurrieren:

Reviewer: ```suggestion
Gibt man eine Funktion an `setState`, so kann man in der Funktion auf den aktuellen Wert des States zugreifen. Da Aufrufe an `setState` gebündelt durchgeführt werden, kannst du Updates verketten und sicher stellen, dass diese aufeinanderfolgend durchgeführt werden, anstatt miteinander zu kollidieren:
```

```jsx

> erhoeheZaehlerCount() {

Reviewer: ```suggestion
incrementCount() {
```

  this.setState((state) => {
    // Wichtig: verwende `state` statt `this.state` beim Update.

>     return {count: state.zaehler + 1}

Reviewer: ```suggestion
    return {count: state.count + 1}
```

  });
}

> behandleEtwas() {

Reviewer: ```suggestion
handleSomething() {
```

>   // Nehmen wir an, `this.state.zaehler` startet bei 0.

Reviewer: ```suggestion
  // Nehmen wir an, `this.state.count` startet bei 0.
```

>   this.erhoeheZaehler();

Reviewer: ```suggestion
  this.incrementCount();
```

>   this.erhoeheZaehler();

Reviewer: ```suggestion
  this.incrementCount();
```

>   this.erhoeheZaehler();

Reviewer: ```suggestion
  this.incrementCount();
```

>   // Wenn du `this.state.zaehler` verwenden würdest, wäre es immer noch 0.

Reviewer: ```suggestion
  // Wenn du `this.state.count` verwenden würdest, wäre es immer noch 0.
```

  // Wenn React aber die Komponente neu rendert, wird es 3 sein.
}
```

[Lerne mehr über `setState`](/docs/react-component.html#setstate)

### Wann ist `setState` asynchron? {#when-is-setstate-asynchronous}

Aktuell ist `setState` in Event Handlern asynchron.

> Das stellt z.B. sicher, dass `Kind` nicht zwei mal gerendert wird, wenn sowohl `Mutter` und `Kind` `setState` während eines Klick-Events aufrufen. Stattdessen "leert" React die Zustandsveränderungen am Ende des Browser-Events. Das führt in größeren Apps zu signifikanten Performance-Gewinnen.

Reviewer: ```suggestion
Das stellt z.B. sicher, dass das `Kind`-Element nicht zwei mal gerendert wird, wenn sowohl `Eltern`-Element und `Kind`-Element `setState` während eines Klick-Events aufrufen. Stattdessen "leert" React die Veränderungen des States am Ende des Browser-Events. Das führt in größeren Apps zu signifikanten Performance-Gewinnen.
```

Das ist ein Implementierungdetail, verlasse dich also nicht direkt darauf. In zukünftigen Versionen wird React Updates standardmäßig in mehr Fällen bündeln.

### Warum updated React `this.state` nicht synchron? {#why-doesnt-react-update-thisstate-synchronously}

Wie im vorherigen Abschnitt erklärt, wartet React absichtlich bis alle Komponenten `setState()` in ihren Event-Handlern aufrufen, bevor neu gerendert wird. Das verbessert die Performance, da unnötiges Rendern vermieden wird.

Du magst dich immer noch fragen, warum React nicht einfach `this.state` sofort ohne neues Rendern verändert.

Die zwei wichtigsten Gründe sind:

> * Das würde die Konsitenz zwischen `props` und `state` brechen, was zu sehr schwer verständlichen Problemen führen würde.

Reviewer: ```suggestion
* Das würde die Konsistenz zwischen `props` und `state` brechen, und zu Problemen führen die sehr schwer zu debuggen sind.
```

* Das würde einige der neuen Features, an denen wir arbeiten, unmöglich machen.

Dieser [GitHub-Kommentar](https://github.com/facebook/react/issues/11527#issuecomment-360199710) nennt spezifische Beispiele.

### Sollte ich eine Zustandsverwaltungsbibliothek wie Redux oder MobX verwenden? {#should-i-use-a-state-management-library-like-redux-or-mobx}

[Vielleicht.](https://redux.js.org/faq/general#when-should-i-use-redux)

> Es ist eine ziemlich gute Idee, React zu verstehen, bevor man weitere Bibliotheken hinzufügt. Du kannst recht komplexe Applikationen auch mit ausschließlich der Zustandsverwaltung von React bauen.

Reviewer: ```suggestion
Es ist eine ziemlich gute Idee, React zu verstehen, bevor man weitere Bibliotheken hinzufügt. Du kannst auch recht komplexe Applikationen nur mit React erstellen.
```
